storage/cloudspanner/tree_storage.go

```diff
@@ -0,0 +1,523 @@
+	ReadUsingIndex(ctx context.Context, table, index string, keys spanner.KeySet, columns []string) *spanner.RowIterator
+	ReadRow(ctx context.Context, table string, key spanner.Key, columns []string) (*spanner.Row, error)
+}
+
+// latestSTH reads and returns the newest STH.
+func (t *treeStorage) latestSTH(ctx context.Context, stx spanRead, treeID int64) (*spannerpb.TreeHead, error) {
+	query := spanner.NewStatement(
+		"SELECT t.TreeID, t.TimestampNanos, t.TreeSize, t.RootHash, t.RootSignature, t.TreeRevision, t.TreeMetadata FROM TreeHeads t" +
+			"   WHERE t.TreeID = @tree_id" +
+			"   ORDER BY t.TimestampNanos DESC " +
+			"   LIMIT 1")
+	query.Params["tree_id"] = treeID
+
+	var th *spannerpb.TreeHead
+	rows := stx.Query(ctx, query)
+	defer rows.Stop()
+	err := rows.Do(func(r *spanner.Row) error {
+		tth := &spannerpb.TreeHead{}
+		var sig, meta []byte
+		if err := r.Columns(&tth.TreeId, &tth.TsNanos, &tth.TreeSize, &tth.RootHash, &sig, &tth.TreeRevision, &meta); err != nil {
+			return err
+		}
+		sigPB := &spannerpb.DigitallySigned{}
+		if err := proto.Unmarshal(sig, sigPB); err != nil {
+			return err
+		}
+		tth.Signature = sigPB
+
+		metaPB := &any.Any{}
+		if err := proto.Unmarshal(meta, metaPB); err != nil {
+			return err
+		}
+		tth.Metadata = metaPB
+
+		th = tth
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	if th == nil {
+		glog.Warningf("no head found for treeID %v", treeID)
+		return nil, storage.ErrTreeNeedsInit
+	}
+	return th, nil
+}
+
+type newCacheFn func(*trillian.Tree) (cache.SubtreeCache, error)
+
+func (t *treeStorage) getTreeAndConfig(ctx context.Context, treeID int64, opts trees.GetOpts) (*trillian.Tree, proto.Message, error) {
+	tree, err := trees.GetTree(ctx, t.admin, treeID, opts)
+	if err != nil {
+		return nil, nil, err
+	}
+	config, err := unmarshalSettings(tree)
+	if err != nil {
+		return nil, nil, err
+	}
+	return tree, config, nil
+}
+
+// begin returns a newly started tree transaction for the specified tree.
+func (t *treeStorage) begin(ctx context.Context, treeID int64, opts trees.GetOpts, newCache newCacheFn, stx spanRead) (*treeTX, error) {
+	tree, config, err := t.getTreeAndConfig(ctx, treeID, opts)
+	if err != nil {
+		return nil, err
+	}
+	cache, err := newCache(tree)
+	if err != nil {
+		return nil, err
+	}
+	treeTX := &treeTX{
+		treeID: treeID,
+		ts:     t,
+		stx:    stx,
+		cache:  cache,
+		config: config,
+	}
+
+	return treeTX, nil
+}
+
+// getLatestRoot populates this TX with the newest tree root visible by this
+// transaction.
+func (t *treeTX) getLatestRoot(ctx context.Context) error {
+	t.getLatestRootOnce.Do(func() {
+		t._currentSTH, t._currentSTHErr = t.ts.latestSTH(ctx, t.stx, t.treeID)
+		if t._currentSTH != nil {
+			t._writeRev = t._currentSTH.TreeRevision + 1
+		}
+	})
+
+	return t._currentSTHErr
+}
+
+// treeTX is a concrete implementation of the Trillian
+// storage.TreeTX interface.
+type treeTX struct {
```

Comment:
Could split this file into tree, log and map specific code, can be done later if worthwhile.

Reply:
the log_storage and (forthcoming) map_storage files are for that :)

Reply:
I'm blaming GitHub for hiding the log_storage diff as it was "too big".

Reply:
:)